### PR TITLE
Validate custom project names based on actual API rules

### PR DESCRIPTION
**Pull request by horenmar**
## Description
I implemented validation of project names based on rules from AWS lambda spec, according to @vlcinsky analysis.

## GitHub Issues
#521

### CODE REVIEW

**Comment by Miserlou** [comment]
This is a great PR!

Before I merge, can I get some clarification on the '_' character. The previous behavior disallowed it, but this change allows it. Is it valid or not. I seem to remember there was some reason we had to remove '_'.

**Comment by vlcinsky** [comment]
The [AWS spec](http://docs.aws.amazon.com/lambda/latest/dg/API_CreateFunction.html#API_CreateFunction_RequestSyntax) clearly allows use of "_" (underscore).

Node.js variant of zappa (serverless) did not allow underscore what was [considered a bug](https://github.com/serverless/serverless/issues/2433) and now they allow it.

I can imagine some restriction on using underscore when the name would be processed by slugify, but as we validate the name after it was possibly slugified, such restriction would not apply to us.

In short: I think underscore shall be allowed.

**Comment by horenmar** [comment]
The regex in `CreateFunction` documentation for function name is `[a-zA-Z0-9-_]+`, which does imply that underscore is allowed.

**Comment by Miserlou** [comment]
Okay, this seems to work, so I'm going to merge it now.

I am worried this will cause some unknown effects that I don't remember, and it will be a bit tricky for some users during upgrade, but it's the right thing to do.
